malaria-ios/Managers/AchievementManager.swift

```diff
@@ -0,0 +1,174 @@
+    
+    property.name = name
+    property.activation = activationMode
+    property.activationValue = value
+    property.initialValue = initialValue
+    
+    CoreDataHelper.sharedInstance.saveContext(context)
+  }
+  
+  /**
+   Defines an achievement with a name, tag and defined properties name that it will use.
+   
+   Example: defineAchievement("Bounty Hunting",
+   tag: "Games",
+   ["Score 10 Points", "Score 50 Points", "Kill Boss"]
+   */
+  
+  func defineAchievement(name: String,
+                         description: String,
+                         propertyNames: [String],
+                         tag: String) {
+    
+    // Check for duplicate achievements.
+    for achievement in self.achievements {
+      if achievement.name == name {
+        return
+      }
+    }
+    
+    let achievement = NSEntityDescription.insertNewObjectForEntityForName("Achievement", inManagedObjectContext: context) as! Achievement
+    
+    // Check whether the required properties for this achievement exists.
+    let matchingProperties: [AchievementProperty]? =
+      properties.filter { propertyNames.contains($0.name!) }
+    
+    guard matchingProperties != nil else {
+      Logger.Warn("There are no matching properties for the achievement.")
+      return
+    }
+    
+    achievement.properties = Set(matchingProperties!)
+    achievement.name = name
+    achievement.tag = tag
+    achievement.desc = description
+    
+    CoreDataHelper.sharedInstance.saveContext(context)
+  }
+  
+  /// Gets the current value of a `AchievementProperty`.
+  
+  func getValue(propertyName: String) -> Int {
+    return properties.filter { $0.name! == propertyName }
+      .first!
+      .currentValue as! Int
+  }
+  
+  /**
+   Sets the a new value for the `AchievementProperty`.
+   It doesn't overwrite the value if the achievementProperty will be activated.
+   */
+  
+  func setValue(propertyName: String, value: Int) {
+    
+    var value = value
+    let property = properties.filter { $0.name == propertyName }.first
+    
+    // Update value only if we achieved a better (lesser or greater, depeding on the case)
+    // value than the original one, not losing the previous achieved value.
+    
+    let currentPropertyValue = property?.currentValue! as! Int
+    
+    switch property!.activation! {
+    case ">":
+      value = value > currentPropertyValue ? value : currentPropertyValue
+    case "<":
+      value = value < currentPropertyValue ? value : currentPropertyValue
+    default: return
+    }
+    
+    property!.currentValue = value;
+  }
+  
+  /// Set the current value for multiple `AchievementProperty`.
+  
+  func setValue(properties: [AchievementProperty], value: Int) {
+    for property in properties {
+      setValue(property.name!, value: getValue(property.name!) + value)
+    }
+  }
+  
+  /**
+   Returns a list of the unlocked achievements.
+   */
+  
+  func checkAchievements() -> [Achievement] {
```

Comment:
Pros and cons of storing the unlocked achivement and managing it? If this method is called multiple times it would a be a good tradeof memory/computing